feat(apollo-usage-report): report referenced field by types

EmrysMyrddin opened the pull request:
Related to YOGA-19
Related to #3580

ardatan commented on packages/plugins/apollo-usage-report/src/reporter.ts:
```diff
@@ -0,0 +1,189 @@
+      try {
+        const abortCtl = new AbortController();
+        this.#logger?.debug(`Sending report (try ${tries}/5)`);
+        const timeout = setTimeout(() => abortCtl.abort(), this.#options.exportTimeout);
```
You can use `AbortSignal.timeout` here, maybe?

EmrysMyrddin replied:
I've never used this api, is there any way to "cancel" this timeout ? Or we just forget about it once the request have been fullfiled ?

ardatan commented on packages/plugins/apollo-usage-report/src/reporter.ts:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    const encodedReport = Report.encode(report).finish();
+    const compressedReport = new ReadableStream({
```
The same readable stream might not be used multiple times. Worth to double-check

EmrysMyrddin replied:
Oh yeah, I must probably do this for each try

ardatan commented on packages/plugins/apollo-usage-report/src/stats.ts:
```diff
@@ -0,0 +1,668 @@
+// Copied from https://github.com/apollographql/apollo-server/blob/8c6579e5b61276b62dc7e30e6fac9a4242e24daa/packages/server/src/plugin/usageReporting/stats.ts
+/* eslint-disable */
+
+import type { NonFtv1ErrorPath } from '@apollo/server-gateway-interface';
```
We can avoid this package entirely by copying this too, no?

EmrysMyrddin replied:
Yes, we can. We discussed it on slack, and we decided that if the Licence allows it, and the dependency is small, it is ok to have it as a dependence.

That's why I've made a copy of the relevant part of the apollo-server code. I don't think it is worth to embed the whole `apollo-server` deps just for a class ^^'

ardatan commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -121,15 +171,17 @@ export function useApolloUsageReport(options: ApolloUsageReportOptions = {}): Pl
-              () => hashSHA256(printSchema(schema), yoga.fetchAPI),
-              id => {
-                schemaId = id;
+            schemaIdSet$ = hashSHA256(printSchema(schema), yoga.fetchAPI)
```
Maybe we can use `printSchemaWithDirectives` instead to print the directives because supergraphs heavily use it.

ardatan commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -55,24 +58,68 @@ type ApolloUsageReportOptions = ApolloInlineTracePluginOptions & {
+  hostname?: string;
+  /**
+   * The OS identification string.
+   * The format is `${os.platform()}, ${os.type()}, ${os.release()}, ${os.arch()})`
```
`navigator.platform` etc can be used in platform independent way, maybe?

EmrysMyrddin replied:
Oh I didn't know node had `navigator` API.
The thing I don't know is if Apollo is parsing this field or not. If it's paring the string, then we can't use `navigator.platform`.

EmrysMyrddin commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -150,12 +206,17 @@ export function useApolloUsageReport(options: ApolloUsageReportOptions = {}): Pl
+              ? usageReportingSignature(result, operationName)
+              : (context.params.query ?? '');
+
+            ctx.referencedFieldsByType = calculateReferencedFieldsByType({
```
I've refactored a little this hook, to properly check if the document is valid or not.
I would like to avoid a try catch that "hides" error here for now, since it could hide potential bugs.

Can you test again please ?

kroupacz replied:
I think the `document` is there, it's just under a property called `documentAST`. See my example above

kroupacz commented on packages/plugins/apollo-usage-report/src/reporter.ts:
```diff
@@ -0,0 +1,186 @@
+      }
+    }
+
+    throw new Error('Failed to send traces after 5 tries');
```
@EmrysMyrddin This error has already popped up once in our NestJS microservice as "unhandled error". I would rather not throw an error here, but just log this state

kroupacz replied:
But now that I think about it, it probably doesn't matter. It's an "Unhandled error" because it's sent inside the `setTimeout()` method and it shouldn't break incoming requests if the endpoint `usage-reporting.api.apollographql.com/api/ingress/traces` is unavailable

kroupacz replied:
Yes, that would probably the be best. Something like that:
```
this.sendReport(schemaId).catch((error) => {
  if (options.onError) {
    options.onError(error);
  } else {
   this.#logger.error(error);
  }
```

kroupacz commented on packages/plugins/apollo-usage-report/src/reporter.ts:
```diff
@@ -0,0 +1,186 @@
+  async sendReport(schemaId: string) {
+    const sending = this.#sendReport(schemaId);
+    this.#sending.push(sending);
+    sending.finally(() => this.#sending?.filter(p => p !== sending));
```
I thought that was your original idea 😉 however this code does not currently clean the `#sending` array. This code does nothing 😄

EmrysMyrddin replied:
🤦‍♂️

EmrysMyrddin commented:
@kroupacz Ok, I've fixed the last things you have reported and added unit tests. I've found a few minor issues along the way :-)
Please take a look and don't hesitate if you see some missing tests !

I've also added the `onValidate` hook to handle both unparsable and invaild queries.

kroupacz commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -55,24 +59,82 @@ type ApolloUsageReportOptions = ApolloInlineTracePluginOptions & {
+   * The class to be used to keep track of traces and send them to the GraphOS endpoint
+   * Note: This option is aimed to be used for testing purposes
+   */
+  reporter?: (
```
@EmrysMyrddin Why are the `Reporter` constructor options exposed to the client? I don't see much reason for this and I don't think it's right.

kroupacz replied:
I would recommend not export this to the client side unless it is really necessary and I would rather mock those "options" via spyOn/replaceProperty

kroupacz replied:
@EmrysMyrddin I personally don't like it because I don't believe that _"users will just not use it"_. 😄  But if this is the usual pattern here, then it's fine.
Could you please release a new alpha version for me to test?

kroupacz replied:
Sorry, I already see the current (`@graphql-yoga/plugin-apollo-usage-report@0.9.0-alpha-20250602093039-da21e7b1322d204bdc7f3b2737561e1c26c2bc17`) version above. I didn't notice that the post above from `github-actions` bot was updated.

kroupacz commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -147,15 +235,38 @@ export function useApolloUsageReport(options: ApolloUsageReportOptions = {}): Pl
-            ctx.operationKey = `# ${operationName || '-'}\n${signature}`;
-            ctx.schemaId = schemaId;
+            if (!isDocumentNode(result)) {
+              ctx.operationKey = `# ${context.params.operationName || '-'} \n${context.params.query ?? ''}`;
```
@EmrysMyrddin What is the use-case of the logic that remained in the onParse hook?

EmrysMyrddin replied:
If there is a parsing error, we probably still want to have a trace for this operation.
And if there is a parse error the `onValidate` is not called.

kroupacz replied:
I tested everything I could think of and everything seems to be ok. 👍 
The only thing I would change is this behavior when the query cannot be parsed or is not valid (logic below). I would add an option to the config to disable sending these metrics like the official `ApolloServerPluginUsageReporting` has => [options.sendUnexecutableOperationDocuments](https://github.com/apollographql/apollo-server/blob/98b1976d86633e0370896fa4d4513ed0689a0d8c/packages/server/src/plugin/usageReporting/plugin.ts#L683)
In case of the parsing or validation fails, I would set the `operationKey` e.g. as follows:
- parsing error: 
```
ctx.operationKey = `# ${context.params.operationName || '#GraphQLParseFailure'} \n${context.params.query ?? ''}`;
```
- validation error: 
```
ctx.operationKey = `# ${context.params.operationName || '#GraphQLValidationFailure'} \n${context.params.query ?? ''}`;
```

This is an example of how the metric looks like in Apollo Studio, which could not be parsed. We personally don't need such metrics. It could easily be some requests from an "unknown public bot"... 😞 
<img width="1904" alt="Snímek obrazovky 2025-06-03 v 16 38 28 (2)" src="https://github.com/user-attachments/assets/8edbbbc6-5d74-4bd9-af0c-90944d32c180" />

EmrysMyrddin replied:
Thank you for pointing this out, I've added some test and changed the implementation to closely follow the Apollo implementation for unexecutable operations.

I hope those operations shows up better in Apollo Studio now :-)

Anyway, I've also added the option to ignore those requests and not send traces for them.

kroupacz commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -55,24 +59,87 @@ type ApolloUsageReportOptions = ApolloInlineTracePluginOptions & {
+   * If false, unexecutable operation (with parsing or validation error) will not be sent
+   * @default true
+   */
+  sendUnexecutableOperationDocuments?: boolean;
```
@EmrysMyrddin Thanks for adding this option. 👍 However, in my opinion it would be better to set the default to `false` so that it doesn't unnecessarily/unexpectedly consume the limits on the number of metrics in Apollo Studio. ApolloServerPluginUsageReporting doesn't send it there by default either 😉

kroupacz commented on packages/plugins/apollo-usage-report/src/index.ts:
```diff
@@ -147,15 +240,58 @@ export function useApolloUsageReport(options: ApolloUsageReportOptions = {}): Pl
+              if (getOperationAST(result, context.params.operationName)) {
+                return;
+              } 
+                ctx.operationKey = `## GraphQLUnknownOperationName\n`;
```
In my opinion, this will never happen and in all cases it will be overridden in the onValidate hook.

EmrysMyrddin replied:
It happens when you provide a `operationName` in the body of your request that doesn't actually exists.

When this happens, the result of parsing phase is a `DocumentNode`, but `onValidate` never gets called. So I had to check for it here.

There is a unit test to verify that this case is covered properly :-)

EmrysMyrddin commented:
@kroupacz I've made the last change of option's default value.
If you are ok with the current state of this PR, I will ask for a review from other guild members and merge this :-)

kroupacz commented:
@EmrysMyrddin I'll deploy the latest version of this plugin to our dev environment and let you know by the end of the day. Otherwise, thanks for the great cooperation. 👍

kroupacz commented on packages/plugins/apollo-usage-report/package.json:
```diff
@@ -41,7 +41,10 @@
+    "@apollo/server-gateway-interface": "^1.1.1",
     "@apollo/usage-reporting-protobuf": "^4.1.1",
+    "@apollo/utils.usagereporting": "^2.1.0",
+    "@graphql-tools/utils": "10.9.0-alpha-20250601230319-2f5f24f393b56915c1c913d3f057531f79991587",
```
@EmrysMyrddin FYI: I just noticed that this PR was merged with this alpha version in `package.json`.
